subxt/src/runtime_api/runtime_payload.rs

```diff
@@ -0,0 +1,186 @@
+    /// The return type of the function call.
+    // Note: `DecodeWithMetadata` is needed to decode the function call result
+    // with the `subxt::Metadata.
+    type ReturnType: DecodeWithMetadata;
+
+    /// The runtime API function name.
+    fn fn_name(&self) -> &str;
+
+    /// Scale encode the arguments data.
+    fn encode_args_to(&self, metadata: &Metadata, out: &mut Vec<u8>) -> Result<(), Error>;
+
+    /// Encode arguments data and return the output. This is a convenience
+    /// wrapper around [`RuntimeApiPayload::encode_args_to`].
+    fn encode_args(&self, metadata: &Metadata) -> Result<Vec<u8>, Error> {
+        let mut v = Vec::new();
+        self.encode_args_to(metadata, &mut v)?;
+        Ok(v)
+    }
+
+    /// Returns the details needed to validate the runtime API call, which
+    /// include a statically generated validation hash.
+    fn validation_details(&self) -> Option<ValidationDetails> {
+        None
+    }
+}
+
+pub struct ValidationDetails {
+    /// A hash (this is generated at compile time in our codegen)
+    /// to compare against the runtime code.
+    pub hash: [u8; 32],
+}
+
+/// A runtime API payload containing the generic argument data
+/// and interpreting the result of the call as `ReturnTy`.
+///
+/// This can be created from static values (ie those generated
+/// via the `subxt` macro) or dynamic values via [`dynamic`].
+#[derive(Clone, Debug)]
+pub struct Payload<ArgsData, ReturnTy> {
+    fn_name: Cow<'static, str>,
+    args_data: ArgsData,
+    validation_hash: Option<[u8; 32]>,
+    _marker: PhantomData<ReturnTy>,
+}
+
+impl<ReturnTy: DecodeWithMetadata> RuntimeApiPayload for StaticRuntimeApiPayload<ReturnTy> {
+    type ReturnType = ReturnTy;
+
+    fn fn_name(&self) -> &str {
+        &self.fn_name
+    }
+
+    fn encode_args_to(&self, _metadata: &Metadata, out: &mut Vec<u8>) -> Result<(), Error> {
+        out.extend(self.args_data.clone());
+        Ok(())
+    }
+
+    fn validation_details(&self) -> Option<ValidationDetails> {
+        self.validation_hash.map(|hash| ValidationDetails { hash })
+    }
+}
+
+impl RuntimeApiPayload for DynamicRuntimeApiPayload {
+    type ReturnType = DecodedValueThunk;
+
+    fn fn_name(&self) -> &str {
+        &self.fn_name
+    }
+
+    fn encode_args_to(&self, metadata: &Metadata, out: &mut Vec<u8>) -> Result<(), Error> {
+        let fn_metadata = metadata.runtime_fn(&self.fn_name)?;
+
+        self.args_data
+            .encode_as_fields_to(fn_metadata.fields(), metadata.types(), out)?;
+
+        Ok(())
+    }
+}
+
+/// A static runtime API payload.
+///
+/// The payload already contains the scale-encoded argument bytes.
+///
+/// This is only expected to be used from codegen.
+pub type StaticRuntimeApiPayload<ReturnTy> = Payload<Vec<u8>, ReturnTy>;
+
+/// A dynamic runtime API payload.
+pub type DynamicRuntimeApiPayload = Payload<Composite<()>, DecodedValueThunk>;
+
+impl<ReturnTy: Decode, ArgsData> Payload<ArgsData, ReturnTy> {
```

Comment:
do we need `Decode` here? I don't think anything should need to rely on this?

And if this goes away, we can merge the below `impl` too so `new()` and `new_static()` can live next to eachother similarto eg `TxPayload` :)

Reply:
Yep, have removed the bounds and merged the code :D